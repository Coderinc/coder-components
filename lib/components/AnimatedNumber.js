import React from 'react';
import raf from 'raf';

class AnimatedNumber extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currentNumber: 0,
    };

    this.confirmSixtyFPS = this.confirmSixtyFPS.bind(this);
    this.prepTween = this.prepTween.bind(this);
    this.endTween = this.endTween.bind(this);
    this.tweenValue = this.tweenValue.bind(this);
  }

  componentDidMount() {
    this.prepTween();
  }

  componentDidUpdate(prevProps) {
    if (this.state.currentNumber === this.props.number) {return;}
    if (prevProps.number === this.props.number) {return;}
    if (this.tweenHandler) {this.endTween();}

    this.prepTween();
  }

  confirmSixtyFPS(timestamp) {
    const { lastTimestamp } = this.state;
    return !lastTimestamp || timestamp - lastTimestamp > 16;
  }

  prepTween() {
    this.tweenHandler = raf(timestamp => {
      this.tweenValue(timestamp, true);
    });
  }

  endTween() {
    raf.cancel(this.tweenHandler);
  }

  tweenValue(timestamp, start = false) {
    if (!this.confirmSixtyFPS(timestamp)) {
      this.tweenHandler = raf(this.tweenValue);
      return;
    }

    const { number, duration } = this.props;
    const { currentNumber } = this.state;
    const startTime = start ? timestamp : this.state.startTime;
    const fromNumber = start ? currentNumber : this.state.fromNumber;

    let newNumber;
    if (timestamp - startTime >= duration) {
      newNumber = number;
    } else {
      newNumber =
        fromNumber +
        (number - fromNumber) * ((timestamp - startTime) / duration);
    }

    if (newNumber === number) {
      this.endTween();
      this.setState({
        currentNumber: number,
      });
      return;
    }

    this.setState({
      currentNumber: newNumber,
      startTime: startTime || timestamp,
      fromNumber,
      lastTimestamp: timestamp,
    });
    this.tweenHandler = raf(this.tweenValue);
  }

  render() {
    const displayValue = this.props.noDecimal
      ? Math.round(this.state.currentNumber)
      : this.state.currentNumber;

    const C = this.props.component;
    const formattedValue = this.props.format
      ? this.props.format(displayValue)
      : displayValue;

    if (C) {
      return <C>{formattedValue}</C>;
    }

    return <span>{formattedValue}</span>;
  }
}

export default AnimatedNumber;
