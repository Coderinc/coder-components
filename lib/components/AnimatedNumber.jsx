import React from 'react';
import PropTypes from 'prop-types';
import raf from 'raf';

class AnimatedNumber extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currentNumber: 0,
    };

    this.confirmSixtyFPS = this.confirmSixtyFPS.bind(this);
    this.prepTween = this.prepTween.bind(this);
    this.endTween = this.endTween.bind(this);
    this.tweenValue = this.tweenValue.bind(this);
  }

  componentDidMount() {
    this.prepTween();
  }

  componentDidUpdate(prevProps) {
    const { currentNumber } = this.state;
    const { number } = this.props;
    if (currentNumber === number) {
      return;
    }
    if (prevProps.number === number) {
      return;
    }
    if (this.tweenHandler) {
      this.endTween();
    }

    this.prepTween();
  }

  confirmSixtyFPS(timestamp) {
    const { lastTimestamp } = this.state;
    return !lastTimestamp || timestamp - lastTimestamp > 16;
  }

  prepTween() {
    this.tweenHandler = raf((timestamp) => {
      this.tweenValue(timestamp, true);
    });
  }

  endTween() {
    raf.cancel(this.tweenHandler);
  }

  tweenValue(timestamp, start = false) {
    if (!this.confirmSixtyFPS(timestamp)) {
      this.tweenHandler = raf(this.tweenValue);
      return;
    }

    const { number, duration } = this.props;
    const { currentNumber, startTime, fromNumber } = this.state;
    const validatedStartTime = start ? timestamp : startTime;
    const validatedFromNumber = start ? currentNumber : fromNumber;

    let newNumber;
    if (timestamp - startTime >= duration) {
      newNumber = number;
    } else {
      newNumber = fromNumber + (number - fromNumber) * ((timestamp - startTime) / duration);
    }

    if (newNumber === number) {
      this.endTween();
      this.setState({
        currentNumber: number,
      });
      return;
    }

    this.setState({
      currentNumber: newNumber,
      startTime: validatedStartTime || timestamp,
      fromNumber: validatedFromNumber,
      lastTimestamp: timestamp,
    });
    this.tweenHandler = raf(this.tweenValue);
  }

  render() {
    const { noDecimal, format, component } = this.props;
    const { currentNumber } = this.state;
    const displayValue = noDecimal ? Math.round(currentNumber) : currentNumber;

    const C = component;
    const formattedValue = format ? format(displayValue) : displayValue;

    if (C) {
      return (
        <C>
          {formattedValue}
        </C>
      );
    }

    return (
      <span>
        {formattedValue}
      </span>
    );
  }
}

AnimatedNumber.propTypes = {
  number: PropTypes.number,
  duration: PropTypes.number,
  noDecimal: PropTypes.bool,
  format: PropTypes.func,
  component: PropTypes.func,
};

AnimatedNumber.defaultProps = {
  number: 0,
  duration: 250,
  noDecimal: false,
  format: undefined,
  component: undefined,
};

export default AnimatedNumber;
