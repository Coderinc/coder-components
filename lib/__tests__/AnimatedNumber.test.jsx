import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import AnimatedNumber from '../components/AnimatedNumber';
import { resolve } from 'url';

describe('AnimatedNumber', () => {
  it('renders correctly', () => {
    const component = renderer.create(<AnimatedNumber number={100} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('starts at 0 and goes to 100', (done) => {
    const component = mount(<AnimatedNumber number={100} duration={250} />);
    let value = component.find('span').text();
    expect(value).toEqual("0");
    (new Promise(resolve => setTimeout(resolve, 500)))
    .then(() => {
      value = component.find('span').text();
      expect(value).toEqual("100");
      done();
    })
  });

  it('changes values', (done) => {
    let value;
    const component = mount(<AnimatedNumber number={100} duration={250} />);
    (new Promise(resolve => setTimeout(resolve, 500)))
    .then(() => {
      value = component.find('span').text();
      expect(value).toEqual("100");
      component.setProps({ number: 1000 });
      expect(value).toEqual("100");
      return new Promise(resolve => setTimeout(resolve, 500));
    })
    .then(() => {
      value = component.find('span').text();
      expect(value).toEqual("1000");
      done();
    });
  });
});