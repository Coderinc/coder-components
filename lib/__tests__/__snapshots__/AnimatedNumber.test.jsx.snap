// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedNumber renders correctly 1`] = `
<span>
  0
</span>
`;
